### PR TITLE
Handle crashing micro simulations

### PR DESCRIPTION
**Pull request by tjwsch**
This is an attempt to solve issue #74.
This cannot be a permanent solution to the handling of crashing simulation. If a simulation crashes during the first time step, it is replaced with data from another random simulation. In the long run, I don't see any other solution than an interpolation (in cases where only one simulation is given and it cannot be replaced by one of different complexity).
Moreover, it has not been tested with real simulations, in parallel, or using spack.

### CODE REVIEW

File: micro_manager/micro_manager.py [hunk 1]
```diff
@@ -71,8 +71,6 @@ def __init__(self, config_file: str) -> None:
             self._rank,
             self._size)
 
-        micro_file_name = self._config.get_micro_file_name()
-
```
Comment (IshaanDesai) [inline micro_manager/micro_manager.py:75]: Is this related to crash handling?

File: micro_manager/micro_manager.py [hunk 2]
```diff
@@ -437,13 +438,45 @@ def _solve_micro_simulations(self, micro_sims_input: list) -> list:
         micro_sims_output = [None] * self._local_number_of_sims
 
         for count, sim in enumerate(self._micro_sims):
-            start_time = time.time()
-            micro_sims_output[count] = sim.solve(micro_sims_input[count], self._dt)
-            end_time = time.time()
 
-            if self._is_micro_solve_time_required:
+            if not self._crashed_sims[count]:
+                try:
```
Comment (IshaanDesai) [inline micro_manager/micro_manager.py:443]: Please add comments on every condition or switch case that you write, so that it is easier to understand what the code is doing.

File: micro_manager/micro_manager.py [hunk 3]
```diff
@@ -437,13 +438,45 @@ def _solve_micro_simulations(self, micro_sims_input: list) -> list:
...
+                    start_time = time.time()
+                    micro_sims_output[count] = sim.solve(
+                        micro_sims_input[count], self._dt)
+                    end_time = time.time()
+                except Exception as error_message:
+                    _, mesh_vertex_coords = self._participant.get_mesh_vertex_ids_and_coordinates(
+                        self._macro_mesh_name)
```
Comment (IshaanDesai) [inline micro_manager/micro_manager.py:450]: Rather than having the API call here, we can just save the coordinates and sample them as and when necessary.

File: micro_manager/micro_manager.py [hunk 4]
```diff
@@ -437,13 +438,45 @@ def _solve_micro_simulations(self, micro_sims_input: list) -> list:
...
+                    self._logger.error(error_message)
+                    micro_sims_output[count] = self._old_micro_sims_output[count]
+                    self._crashed_sims[count] = True
+            else:
+                micro_sims_output[count] = self._old_micro_sims_output[count]
+            if self._is_micro_solve_time_required and not self._crashed_sims[count]:
                 micro_sims_output[count]["micro_sim_time"] = end_time - start_time
 
+            crash_ratio = np.sum(self._crashed_sims) / len(self._crashed_sims)
+            if crash_ratio > 0.2:
```
Comment (IshaanDesai) [inline micro_manager/micro_manager.py:464]: Comments would help here. In the future we could potentially make this a user configurable parameter.

File: micro_manager/micro_manager.py [hunk 5]
```diff
@@ -437,13 +438,45 @@ def _solve_micro_simulations(self, micro_sims_input: list) -> list:
...
+                self._logger.info("More than 20% of the micro simulations on rank {} have crashed. "
+                                  "Exiting simulation.".format(self._rank))
```
Comment (IshaanDesai) [inline micro_manager/micro_manager.py:466]: Lets calculate the percentage number in the comment with the ratio.

**Review by IshaanDesai — CHANGES_REQUESTED** [review]
Good first steps, but we need to do a bit more work before merging this. Please look at the comments below. A general comment is writing comments in the code such that it is readable. I will do another round of reviewing soon.

File: micro_manager/micro_manager.py [hunk 6]
```diff
@@ -437,13 +438,45 @@ def _solve_micro_simulations(self, micro_sims_input: list) -> list:
...
+                sys.exit()
+
+        set_sims = np.where(micro_sims_output)
+        none_mask = np.array([item is None for item in micro_sims_output])
+        unset_sims = np.where(none_mask)[0]
+
+        for unset_sims in unset_sims:
+            self._logger.info("Micro Sim {} has previously not run. "
+                              "It will be replace with the output of the first "
+                              "micro sim that ran {}".format(unset_sims, set_sims[0][0]))
```
Comment (IshaanDesai) [inline micro_manager/micro_manager.py:476]:
```suggestion
            self._logger.info("Micro simulation {} has has crashed in the very first run attempt. "
                              "The output of the first micro sim that ran ({}) will be used as its output.".format(unset_sims, set_sims[0][0]))
```

File: micro_manager/micro_manager.py (hunk 1 above)
Comment (tjwsch) [reply micro_manager/micro_manager.py:75]: No this is a formatting error. I will revert this and open a separate pull request with changes.

**Comment by tjwsch** [comment]
The latest commit adds interpolation. Besides lacking extrapolation, scipy's griddata module also cannot interpolate on linearly dependent data, thus data on the dummy problems cannot be interpolated.  more thought must be put into the issue, to interpolate and extrapolate all cases of simulation crashes, and implementing a custom solution for interpolation is most likely required.

**Comment by IshaanDesai** [comment]
Further discussions led to the idea that implementing an [inverse distance weighing](https://en.wikipedia.org/wiki/Inverse_distance_weighting) would make sense for the interpolation. This would be done with a k-nearest-neighbor method, where the user can set k.

File: micro_manager/interpolation.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+class Interpolation:
+    def __init__(self, logger):
+
+        self._logger = logger
+
+    def get_nearest_neighbor_indices_local(
+        self,
+        all_local_coords,
+        inter_point,
```
Comment (IshaanDesai) [inline micro_manager/interpolation.py:13]: Data type of input arguments are missing.

File: micro_manager/interpolation.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+            inter_point:
+                Coordinates of the point for which the neighbors are to be found.
+            k: int
+            inter_point_is_neighbor: bool, optional
+                Decide whether the interpolation point is considered as its own neighbor.
+                Defaults to False.
+
+        Returns: np.ndarray
+            of indices of the k nearest neighbors.
+        """
```
Comment (IshaanDesai) [inline micro_manager/interpolation.py:32]: Stick to [NumPy style docstrings](https://numpydoc.readthedocs.io/en/latest/format.html).

File: micro_manager/interpolation.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        assert (
+            len(all_local_coords) > k
+        ), "Number of neighbors must be less than the number of all available neighbors."
```
Comment (IshaanDesai) [inline micro_manager/interpolation.py:35]:
```suggestion
        ), "Desired number of neighbors must be less than the number of all available neighbors."
```

Comment (IshaanDesai) [inline micro_manager/interpolation.py:35]:
```suggestion
        ), "Desired number of neighbors must be less than the number of all available neighbors."
```

File: micro_manager/interpolation.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+        Note: It can be chosen whether the point itself is considered as a neighbor or not.
+        Args:
+            all_local_coords: list
+                List of coordinates of all points.
+            inter_point:
+                Coordinates of the point for which the neighbors are to be found.
+            k: int
+            inter_point_is_neighbor: bool, optional
+                Decide whether the interpolation point is considered as its own neighbor.
+                Defaults to False.
```
Comment (IshaanDesai) [inline micro_manager/interpolation.py:28]: What is the reasoning behind providing this option? Does it change the accuracy of the interpolation?

File: micro_manager/interpolation.py [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+            Value at interpolation point.
+        """
+        interpol_val = 0
+        summed_weights = 0
+        for inx in range(len(neighbors)):
+            norm = np.linalg.norm(np.array(neighbors[inx]) - np.array(point)) ** 2
+            if norm < 1e-10:
+                return values[inx]
+            interpol_val += values[inx] / norm
+            summed_weights += 1 / norm
```
Comment (IshaanDesai) [inline micro_manager/interpolation.py:81]: Add brief comments on what is done in each line, as right now it is hard to understand which operations are being done.

File: micro_manager/interpolation.py [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+        Args:
+            neighbors: list
+                Coordinates at which the values are known.
+            point:
+                Coordinates at which the value is to be interpolated.
+            values: list
+                Values at the known coordinates.
+
+        Returns: nd.array
+            Value at interpolation point.
```
Comment (IshaanDesai) [inline micro_manager/interpolation.py:72]: Does the return value have to be of the type of `nd.array`?

File: micro_manager/micro_manager.py [hunk 13]
```diff
@@ -331,11 +351,11 @@ def _initialize(self) -> None:
 
         (
             self._mesh_vertex_ids,
-            mesh_vertex_coords,
+            self._mesh_vertex_coords,
```
Comment (IshaanDesai) [inline micro_manager/micro_manager.py:354]: Is this change related to this PR?

File: micro_manager/interpolation.py (hunk 10 above)
Comment (tjwsch) [reply micro_manager/interpolation.py:28]: If the distance is zero, inverse distance weighting will have a zero division but there is also no reason to interpolate as information at the interpolation point exists. In the case of crash interpolation, this case should never occur as there is no information and that's why interpolation is performed. The option can be removed.

File: micro_manager/micro_manager.py (hunk 13 above)
Comment (tjwsch) [reply micro_manager/micro_manager.py:354]: Yes,  `self._mesh_vertex_coords` is used a few times in the crash handling to log where the simulation crashed and it is used for the interpolation.

File: micro_manager/interpolation.py (hunk 12 above)
Comment (tjwsch) [reply micro_manager/interpolation.py:72]: No, it does not

File: micro_manager/micro_manager.py [hunk 14]
```diff
@@ -606,20 +702,87 @@ def _solve_micro_simulations_with_adaptivity(
...
+                        ] = self._interpolation.inv_dist_weighted_interp(
+                            coord, self._mesh_vertex_coords[unset_sim], key_values
+                        )
+                    micro_sims_output[unset_sim] = output_interpol
+                    if self._is_micro_solve_time_required:
+                        micro_sims_output[unset_sim]["micro_sim_time"] = 0
+                    micro_sims_output[unset_sim]["active_state"] = 1
+                    micro_sims_output[unset_sim][
+                        "active_steps"
+                    ] = self._micro_sims_active_steps[unset_sim]
```
Comment (IshaanDesai) [inline micro_manager/micro_manager.py:785]: We can probably move this into a function and call it once to keep the code clean. Right now we have too many if condition and loop going around, which makes it extremely hard to read code. In general up to 3 levels of loops are fine, more are a problem.

**Review by IshaanDesai — CHANGES_REQUESTED** [review]
Good effort already :+1: my suggestions are mainly about code styling. Let us test this functionality on a real example and check if crashes are caught.

File: micro_manager/micro_manager.py (hunk 13 above)
Comment (IshaanDesai) [reply micro_manager/micro_manager.py:354]: Okay got it :+1:

File: micro_manager/interpolation.py [hunk 15]
```diff
@@ -0,0 +1,84 @@
+import numpy as np
+from sklearn.neighbors import NearestNeighbors
+
+
+class Interpolation:
+    def __init__(self, logger):
+
+        self._logger = logger
+
+    def get_nearest_neighbor_indices_local(
```
Comment (IshaanDesai) [inline micro_manager/interpolation.py:10]:
```suggestion
    def get_nearest_neighbor_indices(
```
Unless we have a global variant of the same function, no need to specify `local` in the function name. We can just go with a generic function name and specify the local indices part in the docstring.

File: micro_manager/interpolation.py [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+        self,
+        all_local_coords: np.ndarray,
+        inter_point: np.ndarray,
+        k: int,
+    ) -> np.ndarray:
+        """
+        Get the indices of the k nearest neighbors of a point in a list of coordinates.
```
Comment (IshaanDesai) [inline micro_manager/interpolation.py:17]:
```suggestion
        Get local indices of the k nearest neighbors of a point.
```

File: micro_manager/interpolation.py [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+
+
+class Interpolation:
+    def __init__(self, logger):
+
+        self._logger = logger
+
+    def get_nearest_neighbor_indices_local(
+        self,
+        all_local_coords: np.ndarray,
```
Comment (IshaanDesai) [inline micro_manager/interpolation.py:12]:
```suggestion
        all_coords: np.ndarray,
```

File: micro_manager/interpolation.py [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+        inter_point: np.ndarray,
+        k: int,
+    ) -> np.ndarray:
+        """
+        Get the indices of the k nearest neighbors of a point in a list of coordinates.
+
+        Parameters
+        ----------
+        all_local_coords : list
```
Comment (IshaanDesai) [inline micro_manager/interpolation.py:21]:
```suggestion
        all_coords : list
```

File: micro_manager/interpolation.py [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+            List of coordinates of all points.
+        inter_point : list | np.ndarray
+            Coordinates of the point for which the neighbors are to be found.
+        k : int
+            Number of neighbors to consider.
+
+        Returns
+        ------
+        neighbor_indices : np.ndarray
+            Indices of the k nearest neighbors in all local points.
```
Comment (IshaanDesai) [inline micro_manager/interpolation.py:31]:
```suggestion
            Local indices of the k nearest neighbors in all local points.
```

File: micro_manager/interpolation.py [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+        """
+        assert len(all_local_coords) > 0, "No local coordinates provided."
```
Comment (IshaanDesai) [inline micro_manager/interpolation.py:33]: Please do a double check if this assertion makes sense here. If we never hit this case in the `micro_manager.py`, we do not need the check. Such checks typically only make sense for user-side API functions.

File: micro_manager/interpolation.py [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+        if len(all_local_coords) < k:
+            self._logger.info(
+                "Number of desired neighbors k = {} is larger than the number of available neighbors {}. Setting k = {}.".format(
```
Comment (IshaanDesai) [inline micro_manager/interpolation.py:36]:
```suggestion
                "Number of desired neighbors k = {} is larger than the number of available neighbors {}. Resetting k = {}.".format(
```

File: micro_manager/interpolation.py [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+
+        neighbor_indices = neighbors.kneighbors(
+            [inter_point], return_distance=False
+        ).flatten()
+
+        return neighbor_indices
+
+    def interpolate(self, neighbors: np.ndarray, point: np.ndarray, values):
+        """
+            Interpolate a value at a point using inverse distance weighting.
```
Comment (IshaanDesai) [inline micro_manager/interpolation.py:51]:
```suggestion
            Interpolate a value at a point using inverse distance weighting. (https://en.wikipedia.org/wiki/Inverse_distance_weighting)
```

File: micro_manager/micro_manager.py [hunk 23]
```diff
@@ -551,12 +580,48 @@ def _solve_micro_simulations(self, micro_sims_input: list) -> list:
...
+                        "Micro simulation at macro coordinates {} has experienced an error. "
+                        "See next entry on this rank for error message.".format(
+                            self._mesh_vertex_coords[count]
+                        )
+                    )
+                    self._logger.error(error_message)
+                    self._has_sim_crashed[count] = True
+
+        # Interpolate result for crashed simulation
+        if self._has_sim_crashed.count(True) > 0:
```
Comment (IshaanDesai) [inline micro_manager/micro_manager.py:610]: Is this condition a must? If there are no crashed simulations, I would expect the logic in the next step to yield `unset_sims = []`.

File: micro_manager/micro_manager.py [hunk 24]
```diff
@@ -654,6 +755,110 @@ def _solve_micro_simulations_with_adaptivity(
...
+            Result of the interpolation in which keys are names of data and the values are the data.
+        """
+        # Find neighbors of the crashed simulation in active and non-crashed simulations
+        # Set iteration length to only iterate over active simulations
+        if self._is_adaptivity_on:
+            iter_length = active_sim_ids
+        else:
+            iter_length = range(len(micro_sims_input))
+        micro_sims_active_input_lists = []  # DECLARATION
+        micro_sims_active_values = []  # DECLARATION
```
Comment (IshaanDesai) [inline micro_manager/micro_manager.py:792]:
```suggestion
        micro_sims_active_input_lists = []
        micro_sims_active_values = []
```
The `DECLARATION` tag is used for member variables and not local variables within a function.

File: tests/unit/test_micro_simulation_crash_handling.py [hunk 25]
```diff
@@ -0,0 +1,125 @@
...
+        return {
+            "micro-vector-data": macro_data["macro-vector-data"],
+            "micro-scalar-data": macro_data["macro-scalar-data"],
+        }
+
+
+class TestSimulationCrashHandling(TestCase):
+    def test_crash_handling(self):
+        """
+        Test if the micro manager catches a simulation crash and handles it adequately.
```
Comment (IshaanDesai) [inline tests/unit/test_micro_simulation_crash_handling.py:29]:
```suggestion
        Test if the Micro Manager catches a simulation crash and handles it adequately.
```

**Review by IshaanDesai — COMMENTED** [review]
We are getting closer to merging this :grin: some more comments on my side. The logic and overall code looks good now.

File: micro_manager/interpolation.py (hunk 20 above)
Comment (tjwsch) [reply micro_manager/interpolation.py:33]: Yes, this is redundant as this is ensured within the micro-manager.

File: micro_manager/micro_manager.py (hunk 23 above)
Comment (tjwsch) [reply micro_manager/micro_manager.py:610]: No, it is redundant.

**Review by IshaanDesai — APPROVED** [review]
Good to go :+1: please remember to **squash and merge**.
